chain/rand/rand.go

```diff
@@ -123,48 +124,51 @@ func NewStateRand(cs *store.ChainStore, blks []cid.Cid, b beacon.Schedule, netwo
 }
 
 // network v0-12
-func (sr *stateRand) getBeaconRandomnessV1(ctx context.Context, round abi.ChainEpoch) ([32]byte, error) {
+func (sr *stateRand) getBeaconEntryV1(ctx context.Context, round abi.ChainEpoch) (*types.BeaconEntry, error) {
 	randTs, err := sr.GetBeaconRandomnessTipset(ctx, round, true)
 	if err != nil {
-		return [32]byte{}, err
-	}
-
-	be, err := sr.cs.GetLatestBeaconEntry(ctx, randTs)
-	if err != nil {
-		return [32]byte{}, err
+		return nil, err
 	}
-
-	return blake2b.Sum256(be.Data), nil
+	return sr.cs.GetLatestBeaconEntry(ctx, randTs)
 }
 
 // network v13
-func (sr *stateRand) getBeaconRandomnessV2(ctx context.Context, round abi.ChainEpoch) ([32]byte, error) {
+func (sr *stateRand) getBeaconEntryV2(ctx context.Context, round abi.ChainEpoch) (*types.BeaconEntry, error) {
 	randTs, err := sr.GetBeaconRandomnessTipset(ctx, round, false)
 	if err != nil {
-		return [32]byte{}, err
+		return nil, err
 	}
+	return sr.cs.GetLatestBeaconEntry(ctx, randTs)
+}
 
-	be, err := sr.cs.GetLatestBeaconEntry(ctx, randTs)
+// network v14 and on
+func (sr *stateRand) getBeaconEntryV3(ctx context.Context, filecoinEpoch abi.ChainEpoch) (*types.BeaconEntry, error) {
+	randTs, err := sr.GetBeaconRandomnessTipset(ctx, filecoinEpoch, false)
 	if err != nil {
-		return [32]byte{}, err
+		return nil, err
 	}
 
-	return blake2b.Sum256(be.Data), nil
-}
+	nv := sr.networkVersionGetter(ctx, filecoinEpoch)
 
-// network v14 and on
-func (sr *stateRand) getBeaconRandomnessV3(ctx context.Context, filecoinEpoch abi.ChainEpoch) ([32]byte, error) {
-	if filecoinEpoch < 0 {
-		return sr.getBeaconRandomnessV2(ctx, filecoinEpoch)
-	}
+	round := sr.beacon.BeaconForEpoch(filecoinEpoch).MaxBeaconRoundForEpoch(nv, filecoinEpoch)
 
-	be, err := sr.extractBeaconEntryForEpoch(ctx, filecoinEpoch)
-	if err != nil {
-		log.Errorf("failed to get beacon entry as expected: %s", err)
-		return [32]byte{}, err
+	for i := 0; i < 20; i++ {
+		cbe := randTs.Blocks()[0].BeaconEntries
+		for _, v := range cbe {
+			if v.Round == round {
+				return &v, nil
+			}
+		}
+
+		next, err := sr.cs.LoadTipSet(ctx, randTs.Parents())
+		if err != nil {
+			return nil, xerrors.Errorf("failed to load parents when searching back for beacon entry: %w", err)
+		}
+
+		randTs = next
```

Comment:
yes, it is weird, and I was tempted to rip out the loop and return an error if it's not found in the next non-null epoch. It came in https://github.com/filecoin-project/lotus/pull/7376 and I suspect it was just a rough copy of the logic in `GetLatestBeaconEntry` which searches 20 tipsets for one with a beacon in it and fails after it goes through 20 without finding a single beacon.

With the current logic, even that search should fail, but historically maybe that was not unreasonable?

Reply:
This logic is for testnets/devnets, which are lower block time networks. In the mainnet, we expect every block to carry a beacon.

Reply:
oh, this is excellent context, so I've added it as a comment in the two places where we walk back 20 tipsets to find beacon entries; it will mystify future maintainers a little less